qutebrowser/browser/webkit/network/pac.py

```diff
@@ -0,0 +1,529 @@
+        """Test is host is a plain hostname.
+
+        Return True if and only if there is no domain name in the hostname (no
+        dots).
+
+        Args:
+            host: The hostname from the URL (excluding port number).
+        """
+        return '.' not in host
+
+    @_js_slot(str, str)
+    def dnsDomainIs(self, host, domain):
+        """Test is host belongs to domain.
+
+        Return true if and only if the domain of hostname matches.
+
+        Args:
+            host: The hostname from the URL.
+            domain: The domain name to test the hostname against.
+        """
+        return host.endswith(domain)
+
+    @_js_slot(str, str)
+    def localHostOrDomainIs(self, host, hostdom):
+        """Test if host is local or belongs to domain.
+
+        Return true if the hostname matches exactly the specified hostname, or
+        if there is no domain name part in the hostname, but the unqualified
+        hostname matches.
+
+        Args:
+            host: The hostname from the URL.
+            hostdom: Fully qualified hostname to match against.
+        """
+        (_, dot_host, _) = host.partition('.')
+        (unqual_hostdom, _, _) = hostdom.partition('.')
+        if dot_host == "":
+            return host == unqual_hostdom
+        else:
+            return host == hostdom
+
+    @_js_slot(str)
+    def isResolvable(self, host):
+        """Test if hostname is resolvable.
+
+        Tries to resolve the hostname. Return true if succeeds.
+
+        Args:
+            host: The hostname from the URL.
+        """
+        return self.dnsResolve(host) != ""
+
+    @_js_slot(str, str, str)
+    def isInNet(self, host, pattern, mask):
+        """Test if host IP is in network.
+
+        Return True if and only if the IP address of the host matches the
+        specified IP address pattern. Pattern and mask specification is done
+        the same way as for SOCKS configuration.
+
+        Args:
+            host: A DNS hostname, or IP address. If a hostname is passed, it
+                  will be resolved into an IP address by this function.
+            pattern: an IP address pattern in the dot-separated format.
+            mask: mask for the IP address pattern informing which parts of
+                  the IP address should be matched against. 0 means ignore,
+                  255 means match.
+        """
+        host_ip = ipaddress.ip_address(host)
+        network = ipaddress.ip_network("{}/{}".format(pattern, mask))
+        return host_ip in network
+
+    @_js_slot(str)
+    def dnsResolve(self, host):
+        """Resolve a DNS hostname.
+
+        Resolves the given DNS hostname into an IP address, and returns it
+        in the dot-separated format as a string.
+
+        Args:
+            host: hostname to resolve.
+        """
+        ips = QHostInfo.fromName(host)
+        if ips.error() != QHostInfo.NoError or len(ips.addresses()) == 0:
+            return ""
```

Comment:
http://serverfault.com/questions/574776/proxy-auto-config-pac-dnsresolve-function-return-values ; tl;dr it's commonly expected for `dnsResolve` to return an empty string in case of an error.